Partition and ExtendableSquareMatrix data structures

kocotom opened the pull request:
I have implemented two data structures:

1.  `Partition`, an efficient representation of set partition with the ability to
    - matching carrier set elements with corresponding blocks in constant time
    - split blocks
    - easily access former blocks which had been split before and iterate through them efficiently
2. `ExtendableSquareMatrix`, an abstract representation of binary relation over a set, a matrix of counters etc. with the ability to
    - assign a value to the matrix cell
    - read a value from the matrix cell
    - extend the `n x n` matrix to the `(n+1) x (n+1)` matrix (if `n < capacity`)
    - implement the exact inner representation of the matrix in various ways and choose the set/get/extend strategies depending on the context 

I have also implemented three concrete representations of the `ExtendableSquareMatrix` which inherit from that structure and implement the virtual methods get, set and extend:

1. `CascadeSquareMatrix` (I'm not sure whether the name is appropriate)
    - the matrix is represented as a single 1D vector which simulates some kind of "cascade" traversal of the matrix in respect to the way the matrix is iteratively extended
    - good for relations/matrices of counters over the sets which are not huge
2. `DynamicSquareMatrix`
    - the matrix is represented as a vector of vectors
3. `HashedSquareMatrix`
    - the matrix is represented as an unordered hashmap
    
These data structures are implemented in the file `partition-relation-pair.hh`. Soon there will be a data structure `PartitionRelationPair` which use combination of `Partition` and `ExtendableSquareMatrix` but it is not fully done yet so I have not included it.

I have also added a lot of tests which work with these data structures.

The code is massively commented to ease understanding.

kocotom committed:
Data structures for partition and extendable square matrix added.
Corresponding tests created.

kocotom committed:
comments modified, mistyped characters fixed

Adda0 commented:
Thank you for the PR. I will be reviewing the PR throughout the week. Is that good enough for your needs?

kocotom committed:
Factory function and copy function for the ExtendableSquaredMatrix added.

kocotom commented:
@Adda0 Thanks for your reaction, yes, that is ok!

I have just made few little changes:

- functions `isReflexive`, `isAntisymetric`, `isTransitive` are now methods of the `ExtendableSquareMatrix` structure, they are not ordinary functions anymore
- I have added the function `create`, it is a factory function which creates an `ExtendableSquareMatrix` instance depending on a given type
- I have added the function `copy`, its purpose is to create a deep copy of `ExtendableSquareMatrix` depending of its type
- I have added few tests

kocotom committed:
Custom copy constructor and operator= added to the Partition structure

kocotom commented:
I have just made one more little change:

- I have implemented custom `copy constructor` for the structure `Partition`
- I have implemented custom `operator=` for the structure `Partition` which preserves information about the capacity of the partition vectors
- I have added few tests to test `copy constructor` and `operator=`

kocotom committed:
Vectors are now passed to functions using const references

kocotom commented:
I have just realized that I had passed several vectors to functions without using constant reference even though these vectors are not modified within corresponding functions, so I have fixed it. Namely:

- `StateBlocks partition = StateBlocks()` changed to `const StateBlocks& partition = StateBlocks()` in `Partition` structure constructor
- `std::vector<State> marked` changed to `const std::vector<State>& marked` in `Partition::splitBlocks` method
- `std::vector<State> states` changed to `const std::vector<State>& states` in `Partition::inSameBlock` method

Adda0 commented:
Hey. Sorry for the delay. Still working on it. Things got in the way.

Adda0 commented on tests/partition-relation-pair.cc:
I noticed that you use dynamic memory for matrices. Is this a design choice? Dynamic memory complicates things, and so far we managed to keep everything without having to manually maintain heap memory allocations.

kocotom replied:
@Adda0 Thanks for your comment! I have decided to use dynamic memory for matrices since I needed to achieve these goals:

- `ExtendableSquareMatrix` is an abstract class which can have several children like `CascadeSquareMatrix`, `DynamicSquareMatrix`, `HashedSquareMatrix` etc.
- it is enough to implement few methods which are virtual in context of `ExtendableSquareMatrix` to create a new concrete type of matrix which chooses its way to represent its contents and manipulate it
- if I create a matrix, I can represent it generally using the pointer type `ExtendableSquareMatrix<T> *` and I do not need to care about the concretely chosen implementation, it is chosen once in the `ExtendableSquareMatrix *create` function

```
template <typename T>
ExtendableSquareMatrix<T> *create(MatrixType type, 
    size_t capacity, size_t size = 0) {
    
    switch(type) {
        case MatrixType::Cascade:
            return new CascadeSquareMatrix<T>(capacity, size);
        case MatrixType::Dynamic:
            return new DynamicSquareMatrix<T>(capacity, size);    
        case MatrixType::Hashed:
            return new HashedSquareMatrix<T>(capacity, size);    
        default:
            return nullptr;
    }
}
```

However, it is not possible to create an instance of the abstract class directly, so I cannot represent the matrix using the type `ExtendableSquareMatrix<T>` in my design. I have to use a pointer or a reference.

I was also thinking about using smart pointers in my design to solve this problem. It could be done easily but I had decided to use manual allocation. Do you suggest preferring smart pointers to manual allocation in context of this library?

Adda0 replied:
Is the handling of the dynamic memory entirely encapsulated in some function, or is it something the user will have to manage themself each time they use the matrices? Furthermore, the automata data structures do not know about these matrices, right?

kocotom replied:
Currently, it works in the following way:

- allocation
    - using the function `ExtendableSquareMatrix<T> *create`, a new matrix will be dynamically allocated on the heap and the corresponding pointer will be returned from the function
    - using the class method `ExtendableSquareMatrix<T> *clone`, a new matrix will be dynamically created on the heap as a deep copy of the given matrix and the corresponding pointer will be returned from the function
- deleting
    - after explicitly calling the `create` or `clone` functions, it is expected that the user will free the memory by using `delete` as soon as the matrix is no longer needed


However, in context of `PartitionRelationPair` class (which will be introduced later and which will use `ExtendableSquareMatrix *` as an attribute), the situation gets more complicated:

- pointer of the type `ExtendableSquareMatrix<char> *` will be a member of the `PartitionRelationPair` class (it will be in fact a matrix of boolean values - a relation)
- there will be two ways of initializing this attribute when creating an instance of `PartitionRelationPair`
    - passing an existing relation (`ExtendableSquareMatrix<char> *`) to the constructor as a pointer. The constructor will dynamically create its own deep copy of the relation on the heap using the `clone` method (or it will use the move semantics and invalidate the former pointer, I am not fully decided yet.)
    - passing a vector of indices `<i_1, i_2, i_3, i_4, ...>` to the constructor which dynamically creates an `ExtendableSquareMatrix` on the heap using the function `create` as a relation `{(i_1, i_2), (i_3, i_4), ...}`
- this matrix could be manipulated only by methods of the `PartitionRelationPair` class
- this matrix will be destroyed automatically in the `~PartitionRelationPair` destructor

You have asked about automata. These matrices will be used in context of computing the simulation relation over automata since I will represent the simulation as a `PartitionRelationPair`. Thus, the automata data structure itself does not know about these matrices. There will be a function which takes an automaton by reference and creates a `PartitionRelationPair` as its simulation relation (or additional functions which use simulation for state space reduction).

To sum up, if the user uses `PartitionRelationPair` class or if they compute simulations, the dynamic memory will be completely encapsulated in the implemented functions and they do not have to care about it. However, if the user creates their own `ExtendableSquareMatrix` for their purposes, they should manually free the matrix. Alternatively, I am able to use smart pointers instead of the current approach.

kocotom replied:
@Adda0 I was thinking about the design of these classes and the classes which I will share later and I have finally decided to use `std::unique_ptr` smart pointers instead of manually handling dynamic memory using `new` and `delete`. Do you think that this approach is more appropriate in context of `mata`?

Adda0 replied:
I would say that unique pointers are great for this kind of use case. Whether they are performant enough remains to be seen.

Adda0 commented on include/mata/utils/partition-relation-pair.hh:
```diff
@@ -0,0 +1,1435 @@
+ *
+ */
+typedef struct Partition
+{
```
Use `class Partition {}` instead of `typedef struct Partition {} Partion;`.

kocotom replied:
Thanks for your suggestion! It should be done by now.

Adda0 commented on include/mata/utils/partition-relation-pair.hh:
```diff
@@ -0,0 +1,1435 @@
+        Partition(const Partition& other);
+
+        // sizes of the used vectors
+        inline size_t numOfStates(void) const { return m_states.size(); }
```
No need to inline methods of a class. They are automatically inlined.

kocotom replied:
Thanks for your suggestion! I have just fixed it.
I have to admit that I did not know that so I am glad you have told me.

kocotom committed:
cloning method added to the ExtendableSquaredMatrix structure and its children

kocotom committed:
missing EOL added

kocotom committed:
Mistyped character

kocotom commented:
I have just did one more little change. I have deleted the `copy` function which had cretead a deep copy of a given `ExtendableSquareMatrix` instance. 
Instead, I have replaced it with a pure virtual method `clone` which is a part of the `ExtendableSquareMatrix` structure. Each substructure reimplements it on its own. I think this is a better solution for this problem.

Adda0 commented on include/mata/utils/partition-relation-pair.hh:
```diff
@@ -0,0 +1,1449 @@
+ *  - remember all ancestors of current blocks and access them
+ *
+ */
+typedef struct Partition
```
Since the class is not a template, its definition could be split into source and header files, which would make reading the file a bit easier.

But considering the other classes are templates, it might be easier to keep `Partition` here as well to keep the file structure simple.

kocotom replied:
Thanks for your comment. I originally decided not to split the class into two files since I had been asked not to create many files. However, I am able to split it if the others agree.

Adda0 commented on include/mata/utils/partition-relation-pair.hh:
```diff
@@ -0,0 +1,1449 @@
+*        vectors of states
+*/
+Partition::Partition(size_t numOfStates, const StateBlocks& partition)
+{
```
I think we have somewhat naturally transitioned in Mata to keeping `{` on the same line as the previous token, instead of inserting a newline between token and the `{`.

kocotom replied:
Thanks for your suggestion! I should have checked the coding style of the library more carefully before creating a pull request. It is fixed by now.

Adda0 commented on include/mata/utils/partition-relation-pair.hh:
```diff
@@ -0,0 +1,1449 @@
+    if(!allStatesUsed)
+    { 
+        m_nodes.push_back({.first = m_states[first], .last = m_states[last]});
+        m_blocks.push_back({.nodeIdx = numOfBlocks-1});
```
Use emplace_back() when constructing in-place. It is faster.

kocotom replied:
Thanks for the suggestion! You are right, I have fixed it.

Adda0 commented on include/mata/utils/partition-relation-pair.hh:
```diff
@@ -0,0 +1,1449 @@
+    // creating a new block and node if there was an unused state
+    if(!allStatesUsed)
+    { 
+        m_nodes.push_back({.first = m_states[first], .last = m_states[last]});
```
Using designated initializers used to cause some issues on either Clang or GCC interchangeably. Let us see whether all of these issues are now fixed in the commonly shipped compilers. If not, it is easy to fix.

kocotom replied:
Thanks for your suggestion. Since I have replaced `push_back` by `emplace_back` when constructing an object/structure in place, I do not use designated initializers anymore.

I have replaced `m_nodes.push_back({.first = m_states[first], .last = m_states[last]});` by `m_nodes.emplace_back(m_states[first], m_states[last]);` and I have implemented the constructor `Node(size_t fst, size_t lst)` to make it work properly. I have also made similar fixes in context of other occurrences.

Adda0 replied:
Sounds good.

Adda0 commented on include/mata/utils/partition-relation-pair.hh:
```diff
@@ -0,0 +1,1449 @@
+ *
+**/
+template <typename T>
+struct CascadeSquareMatrix : public ExtendableSquareMatrix<T>
```
Can be made a class. No need for inlines then.

kocotom replied:
Thanks for your suggestion! I have just changed it (and the other `SquareMatrices` as well).

Adda0 commented on include/mata/utils/partition-relation-pair.hh:
I think it might be more readable if we had all class declarations first, and only then all class method definitions. But it might be OK as it is. Any opinions?

kocotom replied:
Thanks for your suggestion. I am able to do this if the others agree. 
In context of readability, I have to say that there will be one more class in this file in the future (namely `PartitionRelationPair` which will combine `Partition` and `ExtendableSquareMatrix`). It means that there will be several hundred of LOCs added to this file.

Adda0 replied:
Then I would probably split as much as possible, into different files and delimiting declarations from definitions if it is not possible otherwise. But that depends on what others have to say on this topic, too.

kocotom replied:
@Adda0 Thanks for your answer. I have just realized that the current approach can't work. I wrote everything in one header file which contained 

- classes with simple declarations of methods inside of `class { ... }`
- definitions of class methods oustide of `class { ... }`
- inlined ordinary functions

However, I have misunderstood before and the methods were not inlined since they were not defined directly inside of `class { ... }` (I thought that it is sufficient that they are class methods to be inlined and the exact location of their definition does not matter). This caused violation of `ODR`. Thus, I had to change the structure of the file.

Since several of the methods are quite long, I have decided to split the file into multiple files. Currently (after my last commit), the structure is as follows:

- `include/mata/utils/partition.hh` contains declaration of the `Partition` class and methods
- `include/mata/utils/extendable-square-matrix.hh` contains definitions of the `ExtendableSquareMatrix` class and its methods (the same for the subclasses `CascadeSquareMatrix`, `DynamicSquareMatrix`,`HashedSquareMatrix`)
    - The methods are quite short so I kept them inside the `class { ... }` definition and I did not create `extendable-square-matrix.cc` file
- `src/partition.cc` contains definitions of the `Partition` class methods
- `tests/partition.cc` contains tests which cover `Partition` class
- `tests/extendable-square-matrix.cc` contains tests which cover `ExtendableSquareMatrix` class

Is this ok? I was not sure whether the location of the `src/partition.cc` is OK since there is no `src/utils` directory.

Adda0 replied:
I would mirror the folder hierarchy for header files in source as well. Having `src/utils/partition.cc` seems reasonable to me.

Otherwise, I personally am very happy with this separation. Lukáš should have the last say, however.

Adda0 commented on include/mata/utils/partition-relation-pair.hh:
```diff
@@ -0,0 +1,1449 @@
+
+/*************************************
+*
+*        ADDITIONAL FUNCTIONS
```
Not sure here. I think these functions should be inlined, if I am not mistaken.

kocotom replied:
@Adda0 Thanks for your suggestion! I have originally decided not to inline these functions since they

- contain switch statement or several loops
- won't be used frequently 
    - the `operator<<` for `ExtendableSquareMatrix` was primarily designed for debugging purposes
    - the `is_reflexive`, `is_antisymetric` and `is_transitive` functions were primarily designed for `assertions` so any instance of `PartitionRelationPair` where the relation is not a proper partial order cannot be created. If I got it right, the `assertions` won't be ever invoked when compiling with `release`.
    - the `create` function was the only one which I considered for inlining but I did not do it

However, you have mentioned earlier that there is no need to inline any class method. The functions `is_reflexive`, `is_antisymetric` and `is_transitive` belong to the `ExtendableSquareMatrix` class. Could I have misunderstood something?

Adda0 replied:
The way the file is laid out, I thought that these additional functions are functions, not methods. If they were functions with definition in header file, they should be inlined (C++ inline is something different from C inline), otherwise there will be compilation errors during linking when including the file multiple times. See https://softwareengineering.stackexchange.com/a/339490.

kocotom replied:
@Adda0 Thanks for the clarification! The explanation provided in the attached link is really nice. At the moment (after my recent commit):

- the function `operator<<` for the `ExtendableSquareMatrix` class is marked as `inline`
- the function `create` is marked as `inline`
- the function `operator<<` for the `Partition` class is not marked as `inline`. Although it is not a method, it is a friend function of the `Partition` class. If I got it right, even the `friend` functions declared in the class are automatically inlined
- the other provided functions are methods so they are not marked as `inline`

You are right, the placing of `isReflexive, isAntisymetric` and `isTransivite` methods in the file did not make any sense. Initially, I had implemented them as ordinary functions. Then, I decided to rewrite them as class methods, but I did not move them under the corresponding class. I have fixed it in the recent commit.

Adda0 replied:
Looks good to me now. Thanks.

Adda0 commented on include/mata/utils/partition-relation-pair.hh:
```diff
@@ -0,0 +1,1449 @@
+ *
+ * BLOCK ITEMS:
+ * The memory cell m_states[state] contains a corresponding index to the 
+ * 'm_blockItems' vector. The vector 'm_blockItems' has the constant size |S|.
```
In Mata, we agreed to use `member_name_` convention (trailing `_`) to denote member variables/functions. So instead of `m_blockItems`, the member should be called `block_items_`.

We also use `snake_case` instead of `lowerCamelCase`.

kocotom replied:
Thanks for your suggestion! I have committed these changes:

- private attributes `m_varName` renamed to `var_name_` within all classes
- `lowerCamelCase` replaced by `snake_case` in context of 
    - variables
    - functions (which are neither constructors nor destructors)
- `UpperCamelCase` was kept in context of
    - class/structure names
    - aliases created using `using`
    - constructors and destructors
    - elements of the type `enum`
    
Is this ok? If so, the suggestion could be marked as resolved.

Adda0 replied:
This is amazing. Great work.

Adda0 reviewed (commented):
My initial thoughts. I did not review the feasibility of the solution, only the format and the overall structure.

Adda0 commented on include/mata/utils/partition-relation-pair.hh:
```diff
@@ -0,0 +1,1449 @@
+    
+    // this vector contains information about states which has been marked
+    // and helps to detect states which has been marked multiple times
+    std::vector<char> usedStates{};
```
If you use only boolean values, `std::vector<bool>` is space-efficient as it uses bitvectors.

kocotom replied:
@Adda0 Thanks for your suggestion. I tried to avoid using `std::vector<bool>` in this implementation from the very beginning since it does not ensure data locality. Instead, I decided to use `std::vector<char>` even for boolean values even though it is not as space-efficient. However, it is a contiguous block of memory. I discussed it with Lukáš and he agreed.

Adda0 replied:
Sounds good to me. Closing this conversation.

kocotom committed:
Newlines fixed, left curly braces are not on separate lines now

kocotom committed:
lowerCamelCase converted to snake_case. Member variables converted from m_varName to var_name_

kocotom committed:
lowerCamelCase converted to snake_case. Member variables converted from m_varName to var_name_

kocotom committed:
Partition, ExtendableSquareMatrix and its children are now classes (not structures)

kocotom committed:
Class methods are no longer markes as 'inline'

kocotom committed:
push_back replaced by emplace_back when creating object/structure

kocotom commented:
Thanks for your suggestions. I have resolved several of them and I will continue later today. The last fix caused few unsuccessful checks but I have not discovered the cause yet, I will inspect it more thoroughly later today.

kocotom committed:
constructors added to the structures

kocotom committed:
methods get_block_item, get_block, get_node return const reference to the desired element by now

kocotom committed:
Ordinary functions marked as inline, definitions of the methods of the ExtendableSquareMatrix placed after the definition of the class

kocotom committed:
Several tests added to increase the test coverage

kocotom committed:
while cycle replaced by do-while to increase test coverage

kocotom committed:
Several tests added to increase the test coverage

kocotom committed:
Several tests added to increase the test coverage

kocotom committed:
Few tests modified

kocotom commented on include/mata/utils/partition-relation-pair.hh:
```diff
@@ -0,0 +1,1396 @@
+    
+    // creating the initial size and filling the data cells with
+    // default values
+    for(size_t i = 0; i < init_rows; ++i) {extend();}
```
The `codecov` says that this particular line (and two other lines similar to this one) is not fully covered even though I have implemented several tests which cause the following behaviour:

- the cycle is not executed since the condition is not satisfied
- the cycle is executed once
- the cycle is executed multiple times

Any thoughts?

kocotom committed:
Partition and ExtendableSquareMatrix source files/tests split into more files

kocotom committed:
Parentheses added to clarify expression with a bitwise operation

kocotom committed:
Parentheses added to clarify expression with a bitwise operation

kocotom committed:
Parentheses added to clarify expression with a bitwise operation

kocotom committed:
Manual allocation replaced by smart pointers (unique_ptr)

kocotom committed:
[[unlikely]] attribute added, comments fixed

kocotom committed:
Partition BlockItems, Blocks and Nodes reimplemented

kocotom commented:
I have made several changes in the `Partition` class inner logic to make it easier to manipulate with that data structure.

In the previous version, `BlockItem`, `Block` and `Node` were simple structures whose only purpose was to store various indices of partition vectors. The matching of the corresponding structures was done using several methods like `get_node_idx_from_block_item_idx` and so on.

In the current version, `BlockItem`, `Block` and `Node` are private classes defined inside of the `Partition` class. They still store indices of the partition vectors but they also contain methods which simplify manipulation with the partition.

- class `BlockItem`
    - attributes
        - `idx_` - index of the `block_items_` vector corresponding to itself
        - `state_` - corresponding state
        - `block_idx_` - index of the corresponding block
        - `partition_` - const reference to the partition
    - methods
        - `idx` and `state` getters
        - `block` - returns a const reference to the corresponding block
        - `node` - returns a const reference to the corresponding node
        - `repr` - returns a const reference to the corresponding representative

- class `Block`
    - attributes
        - `idx_` - index of the `blocks_` vector corresponding to itself
        - `node_idx_` - index of the corresponding node
        - `partition_` - const reference to the partition
    - methods
        - `idx` getter
        - `node` - returns a const reference to the corresponding node
        - `repr` - returns a const reference to the corresponding representative
        - `first` - returns a const reference to the first corresponding block item
        - `last` - returns a const reference to the last corresponding block item
        - `begin` and `end` - return const iterators which enable using iteration through the block

- class `Node`
    - attributes
        - `idx_` - index of the `nodes_` vector corresponding to itself
        - `first_` - index of the first block item in the node
        - `last_` - index of the last block item in the node
        - `partition_` - const reference to the partition
    - methods
        - `idx` getter
        - `repr` - returns a const reference to the corresponding representative
        - `first` - returns a const reference to the first corresponding block item
        - `last` - returns a const reference to the last corresponding block item
        - `begin` and `end` - return const iterators which enable using iteration through the node

The other methods were changed to reflect the current representation of the partition.

kocotom committed:
extend_and_copy method added to the ExtendableSquareMatrix class

Adda0 commented:
This looks beautiful. I will review the changes during this week. The description above looks great, though.

kocotom committed:
Methods BlockItem::first, BlockItem::last, Node::contains_block added with tests

Adda0 commented on src/partition.cc:
This file should go to `src/utils/partition.cc`. Since we do not have a `src/utils/` folder, I would say we should create one and move this file there.

Adda0 commented on include/mata/utils/extendable-square-matrix.hh:
```diff
@@ -0,0 +1,920 @@
+ *
+**/
+
+using MatrixType = enum MatrixType { None, Cascade, Dynamic, Hashed };
```
Use `enum class`. It is much safer to use (prevents accidental implicit conversions). You use the elements in this enum as elements of an enum class in the code, anyway.

Adda0 commented on include/mata/utils/extendable-square-matrix.hh:
```diff
@@ -0,0 +1,920 @@
+        
+        // type of the matrix which will be chosen as soon as the
+        // child class will be created
+        MatrixType m_type{MatrixType::None};
```
```suggestion
        MatrixType type_{MatrixType::None};
```
If the `m_` stands only for a `member`. We use `<name>_` for private member instead of `m_<name>`.

Adda0 commented on include/mata/utils/extendable-square-matrix.hh:
```diff
@@ -0,0 +1,920 @@
+        * @param[in] j column of the matrix
+        * @param[in] value a value which will be assigned to the memory cell
+        */        
+        virtual void set(size_t i, size_t j, T value = T()) = 0;
```
Default arguments on a virtual method should be prohibited. It a very error-prone feature of C++. See [SO discussion](https://stackoverflow.com/questions/3533589/can-virtual-functions-have-default-parameters), for example.

For the sake of clarity, I think we should split this method into
```cpp
void set(i, j) // Running the default constructor on the type.
void set(i, j, value) // Passing the user-given value by value. 
```

Adda0 commented on include/mata/utils/extendable-square-matrix.hh:
```diff
@@ -0,0 +1,920 @@
+        * @param[in] placeholde value which will be assigned to the 
+        * newly allocated memory cells 
+        */
+        virtual void extend(T placeholder = T()) = 0;
```
The same as with `set()` should be done for `extend()`. So:
```cpp
void extend() // Extending with default initialized values.
void extend(T placeholder) // Extending with a user-specified value.
```

Adda0 commented on include/mata/utils/extendable-square-matrix.hh:
```diff
@@ -0,0 +1,920 @@
+        * the default value of the type T. 
+        * @brief changes the n x n matrix to the (n+1) x (n+1) matrix with
+        * copying of the existing data 
+        * @param[in] placeholde value which will be assigned to the 
```
There is no placeholder in the function arguments. Also, typo.

Adda0 committed:
Rename header guards

Adda0 committed:
Remove `void` in function arguments

Adda0 committed:
Fix typos

Adda0 commented on tests/partition.cc:
Remove debug prints in the tests.

Adda0 reviewed (commented):
The PR looks good to me overall. I have fixed a few typos and simplified some constructs, which will be added in the following commits.

Before merging, all debug prints should be removed from the tests.

When all of these issues are resolved, I think we can merge the PR.

Considering your time constraints, would you like me to resolve these issues so that you can focus on your own problems?

Adda0 commented:
Due to some performance issues with this approach, I am converting the PR to a draft until we decide to work on this PR further. @kocotom, when you have free time again, feel free to let me know how to proceed with this PR further. No rush, though.
